File: 백준/Bronze/2798. 블랙잭/블랙잭.cc
#include <iostream>
#include <stack>
#include <vector>

using namespace std;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int N, M;
    cin >> N >> M;
    int result = 0;
    vector<int> arr(N);

    for (int i = 0; i < N; i++)
    {
        cin >> arr[i];
    }

    for (int i = 0; i < N; i++)
    {
        for (int j = i + 1; j < N; j++)
        {
            for (int k = j + 1; k < N; k++)
            {
                int sum = arr[i] + arr[j] + arr[k];
                if (sum <= M)
                {
                    result = max(result, sum);
                }
            }
        }
    }
    cout << result;
}

File: 백준/Bronze/2798. 블랙잭/블랙잭.java
import java.io.*;
import java.util.StringTokenizer;
import java.util.stream.IntStream;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;
    static int result = 0;
    private static int N;
    private static int M;
    private static int[] arr;

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        arr = new int[N];

        st = new StringTokenizer(br.readLine());
        IntStream.range(0, N).forEach(i -> arr[i] = Integer.parseInt(st.nextToken()));

        combi(0, 0, 0);
        bw.write(String.valueOf(result));
        bw.close();
    }

    static void combi(int start, int depth, int sum) {
        if (depth == 3) {
            if (sum <= M) {
                result = Math.max(result, sum);
            }
            return;
        }
        for (int i = start; i < N; i++) {
            combi(i + 1, depth + 1, sum + arr[i]);
        }
    }
}